internal/cover/cover.go

```diff
@@ -0,0 +1,218 @@
+package cover
+
+import (
+	"fmt"
+	"github.com/benhoyt/goawk/internal/ast"
+	"github.com/benhoyt/goawk/internal/parseutil"
+	. "github.com/benhoyt/goawk/lexer"
+	. "github.com/benhoyt/goawk/parser"
+	"os"
+	"strconv"
+)
+
+const ArrCover = "__COVER"
+
+type coverageHelper struct {
+	covermode     string
+	coverappend   bool
+	fileReader    *parseutil.FileReader
+	annotationIdx int
+	boundaries    map[int]boundary
+	stmtsCnt      map[int]int
+}
+
+type boundary struct {
+	start    Position
+	end      Position
+	fileName string
+}
+
+func New(covermode string, coverappend bool, fileReader *parseutil.FileReader) *coverageHelper {
+	return &coverageHelper{
+		covermode,
+		coverappend,
+		fileReader,
+		0,
+		map[int]boundary{},
+		map[int]int{},
+	}
+}
+
+// Annotate annotates program AST with pieces of coverage tracking code
+func (cov *coverageHelper) Annotate(prog *ast.Program) {
+	prog.Begin = cov.annotateStmtsList(prog.Begin)
+	prog.Actions = cov.annotateActions(prog.Actions)
+	prog.End = cov.annotateStmtsList(prog.End)
+	prog.Functions = cov.annotateFunctions(prog.Functions)
+}
+
+// StoreCoverData writes result coverage report data to coverprofile file
+func (cov *coverageHelper) StoreCoverData(coverprofile string, coverData map[string]string) error {
+	// 1a. If file doesn't exist - create and write covermode line
+	// 1b. If file exists and coverappend=true  - open it for writing in append mode
+	// 1c. If file exists and coverappend=false - truncate it and follow 1a.
+	// 2.  Write all coverData lines
+
+	coverDataInts := prepareCoverData(coverData)
+	isNewFile := true
+
+	var f *os.File
+	if _, err := os.Stat(coverprofile); os.IsNotExist(err) {
+		f, err = os.OpenFile(coverprofile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
+		if err != nil {
+			return err
+		}
+	} else if err == nil {
+		// file exists
+		fileOpt := os.O_TRUNC
+		if cov.coverappend {
+			isNewFile = false
+			fileOpt = os.O_APPEND
+		}
+		f, err = os.OpenFile(coverprofile, os.O_WRONLY|fileOpt, 0644)
+		if err != nil {
+			return err
+		}
+	} else {
+		panic(err)
+	}
+
+	if isNewFile {
+		_, err := f.WriteString("mode: " + cov.covermode + "\n")
+		if err != nil {
+			return err
+		}
+	}
+
+	for i := 1; i <= cov.annotationIdx; i++ {
+		_, err := f.WriteString(renderCoverDataLine(cov.boundaries[i], cov.stmtsCnt[i], coverDataInts[i]))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func prepareCoverData(coverData map[string]string) map[int]int {
+	res := map[int]int{}
+	for k, v := range coverData {
+		ki, err := strconv.Atoi(k)
+		if err != nil {
+			panic("non-int index in coverData: " + k)
```

Comment:
That's true. But in this case it could happen based on user input, in the unlikely case where the user decides to mess with `__COVER` in their AWK program. I'd rather it didn't panic in that case, so let's just return an error like fmt.Errorf(`unexpected non-integer key in __COVER: %q`, k).